File: gpt_tui/services/file_service.py
from __future__ import annotations

import difflib
from pathlib import Path
from typing import Any


class RepoFileService:
    def __init__(self, repo_root: Path) -> None:
        self.repo_root = repo_root.resolve()

    def set_repo_root(self, root: Path) -> tuple[bool, str]:
        path = root.resolve()
        if not path.exists() or not path.is_dir():
            return False, f"Directory not found: {path}"
        self.repo_root = path
        return True, f"Repo root switched to: {self.repo_root}"

    def resolve_repo_path(self, raw_path: str) -> tuple[Path | None, str]:
        if not raw_path:
            return None, "Path is required."
        candidate = Path(raw_path)
        if not candidate.is_absolute():
            return None, "Only absolute paths are allowed."
        resolved = candidate.resolve()
        try:
            resolved.relative_to(self.repo_root)
        except ValueError:
            return None, "Path rejected: outside repo root."
        return resolved, ""

    def list_files(self, root: Path, limit: int = 200) -> list[str]:
        skip_names = {".git", ".gpt-tui", ".venv", "__pycache__", "node_modules"}
        found: list[str] = []
        for path in root.rglob("*"):
            if any(part in skip_names for part in path.parts):
                continue
            if path.is_file():
                found.append(str(path))
            if len(found) >= limit:
                break
        return found

    def snapshot_repo_state(
        self,
        root: Path,
        *,
        limit: int = 2000,
        max_file_bytes: int = 200_000,
    ) -> dict[str, dict[str, Any]]:
        skip_names = {".git", ".gpt-tui", ".venv", "__pycache__", "node_modules"}
        state: dict[str, dict[str, Any]] = {}
        count = 0
        for path in root.rglob("*"):
            if any(part in skip_names for part in path.parts):
                continue
            if not path.is_file():
                continue
            relative = str(path.relative_to(root))
            stat = path.stat()
            text: str | None = None
            if stat.st_size <= max_file_bytes:
                try:
                    text = path.read_text(encoding="utf-8")
                except (OSError, UnicodeDecodeError):
                    text = None
            state[relative] = {
                "path": str(path),
                "size": stat.st_size,
                "mtime_ns": stat.st_mtime_ns,
                "text": text,
            }
            count += 1
            if count >= limit:
                break
        return state

    def diff_repo_state(
        self,
        before: dict[str, dict[str, Any]],
        after: dict[str, dict[str, Any]],
    ) -> list[dict[str, Any]]:
        changes: list[dict[str, Any]] = []
        for relative in sorted(set(before) | set(after)):
            previous = before.get(relative)
            current = after.get(relative)
            if previous is None and current is not None:
                changes.append(
                    {
                        "action": "create",
                        "path": relative,
                        "oldPath": "",
                        "newPath": "",
                        "diff": self._build_diff(relative, "", current.get("text") or ""),
                    }
                )
                continue
            if previous is not None and current is None:
                changes.append(
                    {
                        "action": "delete",
                        "path": relative,
                        "oldPath": "",
                        "newPath": "",
                        "diff": self._build_diff(relative, previous.get("text") or "", ""),
                    }
                )
                continue
            if previous is None or current is None:
                continue
            if (
                previous.get("size") == current.get("size")
                and previous.get("mtime_ns") == current.get("mtime_ns")
            ):
                continue
            before_text = previous.get("text")
            after_text = current.get("text")
            if before_text == after_text and before_text is not None:
                continue
            changes.append(
                {
                    "action": "edit",
                    "path": relative,
                    "oldPath": "",
                    "newPath": "",
                    "diff": self._build_diff(relative, before_text or "", after_text or ""),
                }
            )
        return changes

    def _build_diff(self, relative: str, before: str, after: str) -> str:
        diff = difflib.unified_diff(
            before.splitlines(),
            after.splitlines(),
            fromfile=f"a/{relative}",
            tofile=f"b/{relative}",
            lineterm="",
        )
        return "\n".join(diff)

    def read_utf8(self, file_path: Path, max_chars: int = 6000) -> tuple[str, bool]:
        content = file_path.read_text(encoding="utf-8")
        if len(content) <= max_chars:
            return content, False
        return content[:max_chars] + "\n\n...[truncated]...", True

File: desktop_app/backend/fileService.js
const fs = require("node:fs");
const path = require("node:path");

const SKIP_NAMES = new Set([".git", ".gpt-tui", ".venv", "__pycache__", "node_modules", "out"]);

function walkFiles(root, limit = 200) {
  const found = [];
  const stack = [root];
  while (stack.length && found.length < limit) {
    const current = stack.pop();
    const entries = fs.readdirSync(current, { withFileTypes: true });
    for (const entry of entries) {
      if (SKIP_NAMES.has(entry.name)) {
        continue;
      }
      const fullPath = path.join(current, entry.name);
      if (entry.isDirectory()) {
        stack.push(fullPath);
        continue;
      }
      if (entry.isFile()) {
        found.push(fullPath);
      }
      if (found.length >= limit) {
        break;
      }
    }
  }
  return found;
}

function buildUnifiedDiff(relativePath, beforeText, afterText) {
  const beforeLines = String(beforeText || "").split("\n");
  const afterLines = String(afterText || "").split("\n");
  const header = [`--- a/${relativePath}`, `+++ b/${relativePath}`];
  if (beforeText === afterText) {
    return header.join("\n");
  }

  const body = [];
  const maxLines = Math.max(beforeLines.length, afterLines.length);
  body.push(`@@ -1,${beforeLines.length} +1,${afterLines.length} @@`);
  for (let index = 0; index < maxLines; index += 1) {
    const beforeLine = beforeLines[index];
    const afterLine = afterLines[index];
    if (beforeLine === afterLine) {
      if (beforeLine !== undefined) {
        body.push(` ${beforeLine}`);
      }
      continue;
    }
    if (beforeLine !== undefined) {
      body.push(`-${beforeLine}`);
    }
    if (afterLine !== undefined) {
      body.push(`+${afterLine}`);
    }
  }
  return [...header, ...body].join("\n");
}

class RepoFileService {
  constructor(repoRoot) {
    this.repoRoot = path.resolve(repoRoot);
  }

  setRepoRoot(root) {
    const resolved = path.resolve(root);
    if (!fs.existsSync(resolved) || !fs.statSync(resolved).isDirectory()) {
      return [false, `Directory not found: ${resolved}`];
    }
    this.repoRoot = resolved;
    return [true, `Repo root switched to: ${this.repoRoot}`];
  }

  resolveRepoPath(rawPath) {
    if (!rawPath) {
      return [null, "Path is required."];
    }
    const candidate = path.isAbsolute(rawPath) ? rawPath : path.join(this.repoRoot, rawPath);
    const resolved = path.resolve(candidate);
    const relative = path.relative(this.repoRoot, resolved);
    if (relative.startsWith("..") || path.isAbsolute(relative)) {
      return [null, "Path rejected: outside repo root."];
    }
    return [resolved, ""];
  }

  listFiles(root = this.repoRoot, limit = 200) {
    return walkFiles(path.resolve(root), limit);
  }

  snapshotRepoState(root = this.repoRoot, { limit = 2000, maxFileBytes = 200_000 } = {}) {
    const state = {};
    for (const fullPath of walkFiles(path.resolve(root), limit)) {
      const stats = fs.statSync(fullPath);
      const relativePath = path.relative(root, fullPath).replaceAll("\\", "/");
      let text = null;
      if (stats.size <= maxFileBytes) {
        try {
          text = fs.readFileSync(fullPath, "utf8");
        } catch {
          text = null;
        }
      }
      state[relativePath] = {
        path: fullPath,
        size: stats.size,
        mtimeMs: stats.mtimeMs,
        text,
      };
    }
    return state;
  }

  diffRepoState(before, after) {
    const changes = [];
    const allPaths = new Set([...Object.keys(before), ...Object.keys(after)]);
    for (const relativePath of [...allPaths].sort()) {
      const previous = before[relativePath];
      const current = after[relativePath];
      if (!previous && current) {
        changes.push({
          action: "create",
          path: relativePath,
          oldPath: "",
          newPath: "",
          diff: buildUnifiedDiff(relativePath, "", current.text || ""),
        });
        continue;
      }
      if (previous && !current) {
        changes.push({
          action: "delete",
          path: relativePath,
          oldPath: "",
          newPath: "",
          diff: buildUnifiedDiff(relativePath, previous.text || "", ""),
        });
        continue;
      }
      if (!previous || !current) {
        continue;
      }
      if (previous.size === current.size && previous.mtimeMs === current.mtimeMs) {
        continue;
      }
      const beforeText = previous.text || "";
      const afterText = current.text || "";
      if (beforeText === afterText) {
        continue;
      }
      changes.push({
        action: "edit",
        path: relativePath,
        oldPath: "",
        newPath: "",
        diff: buildUnifiedDiff(relativePath, beforeText, afterText),
      });
    }
    return changes;
  }

  readUtf8(filePath, maxChars = 6000) {
    const content = fs.readFileSync(filePath, "utf8");
    if (content.length <= maxChars) {
      return [content, false];
    }
    return [`${content.slice(0, maxChars)}\n\n...[truncated]...`, true];
  }
}

module.exports = {
  RepoFileService,
  buildUnifiedDiff,
};
